### PR TITLE
adding village events & possible more

### PR DESCRIPTION
**Pull request by Cheaterpaul**
enable faction support for village capture

-todo looking for more event possibilities

### CODE REVIEW

File: src/api/java/de/teamlapen/vampirism/api/event/ModEventFactory.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package de.teamlapen.vampirism.api.event;
+
+import de.teamlapen.vampirism.api.entity.factions.IFaction;
+import de.teamlapen.vampirism.api.entity.factions.IPlayableFaction;
+import de.teamlapen.vampirism.api.world.IVampirismVillage;
+import de.teamlapen.vampirism.world.villages.VampirismVillage;
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/ModEventFactory.java]: Vampirism's classes are not accessible from the API

File: src/api/java/de/teamlapen/vampirism/api/event/ModEventFactory.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import net.minecraftforge.fml.common.eventhandler.Event.Result;
+
+import java.util.List;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class ModEventFactory {
+
+    public static boolean fireVillagerCaptureEvent(@Nullable IVampirismVillage village, @Nonnull List<EntityVillager> villagerIn, @Nullable IPlayableFaction<?> controllingFactionIn, @Nonnull IPlayableFaction<?> capturingFactionIn) {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/ModEventFactory.java]: We don't really need a dedicated ModEventFactory. Forge only does so to minimize the patches to vanilla code.
However, we can keep it, but move it into the main code package (e.g. util), no reason to have this in the api package.

Probably want to move the few other events Vampirism has to this class as well

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 3]
```diff
@@ -45,12 +54,14 @@ public IVampirismVillage getVillage() {
         EntityVillager seedVillager;
         private @Nullable
         EntityVillager newVillager;
-        private boolean willBeVampire;
+        private boolean willBeConverted;
+        private final IPlayableFaction<?> faction;
 
-        public SpawnNewVillager(@Nonnull IVampirismVillage village, @Nonnull EntityVillager seedVillager, boolean willBeVampire) {
+        public SpawnNewVillager(@Nonnull IVampirismVillage village, @Nonnull EntityVillager seedVillager, boolean willBeConverted, @Nonnull IPlayableFaction<?> faction) {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: Does the faction have to be non null?
Maybe we want to spawn a villager in a neutral village.
Also, if it is annotated Nonnull, the getter (and the field) should probably be annotated as well

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 4]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+     * if result is {@link Result#DENY} the Vanilla code is skipped
+     */
+    @HasResult
+    public static class VillagerCaptureFinish extends VampirismVillageEvent {
+
+        private final @Nonnull List<EntityVillager> villager;
+        private final @Nullable IPlayableFaction<?> controllingFaction;
+        private final @Nonnull IPlayableFaction<?> capturingFaction;
+
+        public VillagerCaptureFinish(@Nullable IVampirismVillage village, @Nonnull List<EntityVillager> villagerIn, @Nullable IPlayableFaction<?> controllingFactionIn, @Nonnull IPlayableFaction<?> capturingFactionIn) {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: Probably village should be nonnull here, posting this event for a non existent village does not make much sense

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 5]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+            super(village);
+            villager = villagerIn;
+            controllingFaction = controllingFactionIn;
+            capturingFaction = capturingFactionIn;
+        }
+
+        @Nonnull
+        public List<EntityVillager> getVillager() {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java:181]: Maybe add a comment what these mysterious villagers are

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 6]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+        public IPlayableFaction<?> getCapturingFaction() {
+            return capturingFaction;
+        }
+
+    }
+
+    /**
+     * Fired when a new Capture Entity should be spawned and the faction of the entity is not a Hunter or Vampire Entity
+     */
+    public static class CaptureEntity extends VampirismVillageEvent {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: Maybe "SpawnCaptureEntity"

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 7]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+        public void setPoisonousBlood(boolean poisonous) {
+            poisonousBlood = poisonous;
+        }
+
+    }
+
+    /**
+     * fired when blocks around a village should be replaced (e.g. cursed earth for vampires)
+     */
+    public static class ReplaceBlock extends VampirismVillageEvent {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java:323]: What does this do? 
Cannot be canceled and the block cannot be changed either

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 8]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+     * fired when the caption process is started
+     */
+    @HasResult
+    public static class InitiateCapture extends VampirismVillageEvent {
+
+        private final @Nonnull World world;
+        private final @Nullable IPlayableFaction<?> controllingFaction;
+        private final @Nonnull IPlayableFaction<?> capturingFaction;
+
+        public InitiateCapture(@Nullable IVampirismVillage village, @Nonnull World world, @Nullable IPlayableFaction<?> controllingFaction, @Nonnull IPlayableFaction<?> capturingFaction) {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: Same here nonnull village

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 9]
```diff
@@ -135,4 +153,244 @@ public void setAgressiveVillager(@Nullable IVillageCaptureEntity captureVillager
...
+        public IPlayableFaction<?> getCapturingFaction() {
+            return capturingFaction;
+        }
+
+    }
+
+    /**
+     * fired when the village area is updated (used for vampire fog rendering & sundamage)
+     */
+    public static class RegisterVillageBoundingBoxEvent extends VampirismVillageEvent {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: Maybe "UpdateBoundingBox"

**Review by maxanier — CHANGES_REQUESTED** [review]
Cool. I see you are having big plans for werewolf villages :)

**Comment by maxanier** [comment]
Probably gonna do the final release next week, so anything new you want to add here or can this be "ready for review"?

**Comment by Cheaterpaul** [comment]
when i fixed the code and you are satisfied, it should be finished

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java (hunk 3 above)
Comment (Cheaterpaul) [reply src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java]: until now there is no "village live" without a controlling faction, but now the event will accept it

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java (hunk 7 above)
Comment (maxanier) [reply src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java:323]: Last remaining issue

**Pull request merged**